### ISSUE TITLE
Local Additions after switching from SR to RO back and forth and wiping the database

### ISSUE THREAD

**Issue by tomasz1986** [syncthing/syncthing#7370]
I am not really sure whether all these steps are necessary, but this is how it goes.

1. Share an empty SR folder between Device A and B.
2. Add some files to the folder on Device A and let them sync.
3. Set the folder to RO on Device B.
4. Everything is "Up to Date".

![1](https://user-images.githubusercontent.com/5626656/108008096-dd6d1c80-7042-11eb-8ae3-147f14e723ae.png)
![2](https://user-images.githubusercontent.com/5626656/108008101-df36e000-7042-11eb-9608-21af8a6faec8.png)

5. Shut down Syncthing on Device B, reset the database, and restart.
6. Change the folder to SR on Device B.
7. Everything is "Up to Date".

![3](https://user-images.githubusercontent.com/5626656/108008108-e2ca6700-7042-11eb-8ac3-6b19671cb711.png)
![4](https://user-images.githubusercontent.com/5626656/108008110-e3fb9400-7042-11eb-9b20-7291ed02220e.png)

8. Set the folder to RO on Device A.
9. Shut down Syncthing on Device A, reset the database, and restart.
10. Device A is now marked "Local Additions", even though the files are exactly the same.

![5](https://user-images.githubusercontent.com/5626656/108008114-e5c55780-7042-11eb-8f48-e03e5a185ec6.png)
![6](https://user-images.githubusercontent.com/5626656/108008115-e6f68480-7042-11eb-9b24-66e753286353.png)

Wiping the database on Device A one more time fixes the problem, but why are the files marked "Local Additions" in the first place though?

Logs with `STTRACE=db,model`:

[Device 1.log](https://github.com/syncthing/syncthing/files/5985272/Device.1.log)
[Device 2.log](https://github.com/syncthing/syncthing/files/5985273/Device.2.log)

Databases:

[Device 1 index-v0.14.0.db.zip](https://github.com/syncthing/syncthing/files/5985277/Device.1.index-v0.14.0.db.zip)
[Device 2 index-v0.14.0.db.zip](https://github.com/syncthing/syncthing/files/5985280/Device.2.index-v0.14.0.db.zip)

**Edit**: Just one more thing, as I forgot to add it and do not feel like testing everything again, but in step 9, if you change the folder type from RO to SR on Device A, I believe that this actually leads to conflicts, i.e. the 3 files will be re-downloaded as conflicts, even though if you compare them, they are 100% identical.

**Comment by AudriusButkevicius** [comment]
This is expected, the version vectors need to converge for them to be not local additions, regardless that the files are identical.

**Comment by tomasz1986** [comment]
Hmm, but what is the difference between steps 5-7 and steps 8-10? In both cases, the folder is set to RO before wiping the database. In the former, everything gets marked as "Up to Date", in the latter we get "Local Additions".

**Comment by AudriusButkevicius** [comment]
The devices both need to be send-receive for them to converge on vector clocks, otherwise you end up with the clocks only merged on one side, leading to local additions/remote additions.

In a receive only situation, if you reset the database, you should also remove the data for it to re-download the data and "accept" the remote version of the clocks (and not create it's own version of the clock via scanning before it receives remote data) leading to this problem.

Not sure why this got reopened, I think this is expected hence I'll close it back, and others can reopen if I am wrong.

**Comment by tomasz1986** [comment]
I have done a little bit more testing, and what you are saying seems to be right. Although I must admit that it would be really nice in this situation if Syncthing was able to simply re-use the already existing files instead of re-downloading everything (after pressing the Revert button).

**Comment by AudriusButkevicius** [comment]
You can already do that, set the folder to send-receive, let them converge, set it back to send-only/receive-only.

I mean, it would equality be nice if people stopped doing highly destructive actions and expect things to work without strange states.

**Comment by tomasz1986** [comment]
> You can already do that, set the folder to send-receive, let them converge, set it back to send-only/receive-only.

Yeah, if you are 100% sure that the files are equal. If there is a mix of different files, then pressing Revert is safer, but this results in all the files being deleted and re-downloaded.

> I mean, it would equality be nice if people stopped doing highly destructive actions and expect things to work without strange states.

Of course, but there are cases where there is no choice, e.g. we had to this operation after a disk failure, which corrupted the database, so resetting it was the only solution. We did use the "Revert" button at the end, for what it is worth it.

**Comment by AudriusButkevicius** [comment]
I would have just let it sync from scratch, especially if you are not sure if the data is intact or not. I guess revert changes is that effectively.

You could still achieve what you want.
I would just remove the existing folder, create a new folder with a different id pointing at the old location, shared with nobody, let it scan so it builds the list of stuff it has.

Then add a new folder with the old id pointing at a different location, and let it effectively reuse the data from the old location (between folders) oppose to retransferring.

**Comment by tomasz1986** [comment]
> You could still achieve what you want.
> I would just remove the existing folder, create a new folder with a different id pointing at the old location, shared with nobody, let it scan so it builds the list of stuff it has.
> 
> Then add a new folder with the old id pointing at a different location, and let it effectively reuse the data from the old location (between folders) oppose to retransferring.

This is a really good tip. I have not thought about doing trickery like that.

I guess that next time I will first rename the old folder, add the renamed location as a new "unshared" folder to Syncthing, and then re-add the old one pointing at the original location. Once everything gets "synced" (or really copied), I will remove the old, renamed folder. The only possibly problem may arise if there is not enough space left to store two sets of files at once (although here we did have a lot of free space left).

**Comment by AudriusButkevicius** [comment]
This is how people reuse blocks from their .stversions (in case a file is deleted and then resurrected), by adding an unshared folder pointing at the versions directory.

**Comment by tomasz1986** [comment]
Yeah, I know very well about reusing blocks from other folders, but for some reason I did not even consider doing that in this situation. Although, to be fair, I also did not know that Syncthing was going to re-download everything after pressing the Revert button.

How about this approach?

1. The old folder is still stuck with Local Additons.
2. Move only the files to a different location.
3. Add the new location as a new unshared folder.
4. Revert Local Additions in the old folder.

It requires fewer steps, but will Syncthing also re-use files from the newly added unshared folder rather than downloading them again?
